File: src/main/java/graveler/util/PointedWorld.java
package graveler.util;

import graveler.StressMap;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.world.level.LevelAccessor;
import net.minecraft.world.level.block.AbstractCauldronBlock;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.Blocks;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.chunk.ChunkAccess;

public record PointedWorld(LevelAccessor level, BlockPos pos) {
  private static final int MAX_STRESS = 50;

  public ChunkAccess chunk() {
    return level.getChunk(pos);
  }

  public PointedWorld move(Direction direction) {
    return new PointedWorld(level, pos.relative(direction));
  }

  public BlockState blockState() {
    return level.getBlockState(pos);
  }

  public Block block() {
    return blockState().getBlock();
  }

  public boolean isStressAware() {
    return !blockState().isAir() && !isPermanentlyStable();
  }

  public boolean isPermanentlyStable() {
    return block() == Blocks.BEDROCK || blockState().liquid();
  }

  public boolean allowsFallThrough() {
    return (!blockState().blocksMotion() || blockState().liquid())
        && !(block() instanceof AbstractCauldronBlock);
  }

  public int stress() {
    if (isStressAware()) {
      ChunkAccess chunk = chunk();
      if (chunk == null) {
        return 0;
      }

      StressMap stressMap = WorldExtensions.getStressMap(chunk);
      Integer value = stressMap.get(pos);
      return value == null ? 0 : value;
    }

    if (isPermanentlyStable()) {
      return 0;
    }

    return MAX_STRESS;
  }

  public void setStress(int value) {
    if (!isStressAware()) {
      return;
    }

    ChunkAccess chunk = chunk();
    if (chunk == null) {
      return;
    }

    StressMap stressMap = WorldExtensions.getStressMap(chunk);
    stressMap.set(pos, value);
    chunk.markUnsaved();
  }

  public void clearStress() {
    ChunkAccess chunk = chunk();
    if (chunk == null) {
      return;
    }

    StressMap stressMap = WorldExtensions.getStressMap(chunk);
    if (stressMap.removeIfPresent(pos)) {
      chunk.markUnsaved();
    }
  }
}

File: src/main/kotlin/graveler/PointedWorld.kt
package graveler

import graveler.util.stressMap
import net.minecraft.block.*
import net.minecraft.block.material.Material
import net.minecraft.util.Direction
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import net.minecraft.world.chunk.Chunk

data class PointedWorld(val world: World, val pos: BlockPos) {
  fun getNewStress(): Int {
    if (!isStressAware) return 0

    val below = move(Direction.DOWN)
    val horizontalStress = Direction.Plane.HORIZONTAL
      .map { move(it).stress + 1 }

    return (horizontalStress + below.stress).min()!!
  }

  fun clearStress() {
    chunk?.stressMap?.stresses?.remove(pos)
  }

  var stress: Int
    get() = when {
      isStressAware -> chunk?.stressMap?.get(pos) ?: 0
      isPermanentlyStable -> 0
      else -> MAX_STRESS
    }
    set(value) {
      when {
        isStressAware -> chunk?.stressMap?.set(pos, value)
      }
    }

  private val chunk: Chunk?
    get() = world.getChunkAt(pos)

  val isStressAware: Boolean
    get() = !blockState.isAir(world, pos) && !isPermanentlyStable

  private val isPermanentlyStable: Boolean
    get() = block == Blocks.BEDROCK || material.isLiquid

  fun move(direction: Direction): PointedWorld = copy(pos = pos.offset(direction))

  private val blockState: BlockState
    get() = world.getBlockState(pos)

  private val block: Block
    get() = blockState.block

  private val material: Material
    get() = blockState.material

  val allowsFallThrough: Boolean
    get() = (!material.blocksMovement() || material.isLiquid) &&
      block !is CauldronBlock

  companion object {
    private const val MAX_STRESS = 50
  }
}
